### DISCUSSION TITLE
More in depth upgrade system.

### DISCUSSION THREAD

**Discussion by HeatedWafflez** [FakeFishGames/Barotrauma#5049] [Ideas and suggestions]
Currently the shipyard upgrade system is a fairly nice feature that allows players to improve their ship in areas they see fit. This system although convenient for players wishing to balance out the cons in the ship they're using is fairly linear in the sense that upgrades tend to be subtle. This subtlety means that the ship tends to stay identical throughout the campaign and as a result it's hard to have a ship that feels personalized to the crew using it since it's hard to see any noticeable changes. 

An idea proposed several times before is having purchasable modules for submarines. Examples of this could include...

- Being able to purchase an upgrade that converts the Typhon 1 crew quarters into a ballast.
- Removing the Dugong's electrical discharge coil and instead making it a purchasable upgrade.
- Being able to purchase storage for the Remora's fabrication room since it lacks any and as a result is difficult to use.
- Potentially being able to purchase an extra coilgun for the Orca so that it covers the blind spot beneath it.
- Merging the Humpback's front ballast into one hull and giving it a bigger pump.
- Improved backup power on all vanilla submarines, since they often have multiple batteries feeding into a single relay capped at 1000 Kw/h which effectively renders their backup array useless in most cases.

This idea would greatly benefit both the average player and submarine builders. Submarine builders could designate sections of their ship that must be purchased through a shipyard with the cost also being set by them. Builders could also potentially have it so that two modules conflict with each other meaning crews will have to pick between the upgrade they feel is best for them. An example of this could be...

- Extending the Dugong's medical compartment or increasing the amount of diving suit lockers. The diving suit lockers would be initially capped at either three or two and players could choose between increasing this to five or increasing the size of the medical compartment so that it includes a medical fabricator/bed.

Dugong without any module upgrades.
![dugongnoupgrades](https://user-images.githubusercontent.com/51425624/107839608-3b131780-6d6a-11eb-9b6d-ed0fc42a0c60.png)

Dugong with extra diving suits.
![dugongdivingsuitupgrade](https://user-images.githubusercontent.com/51425624/107839619-50884180-6d6a-11eb-91ff-34f4ad122a16.png)

Dugong with expanded medical compartment.
![dugongbedupgrade](https://user-images.githubusercontent.com/51425624/107839648-84fbfd80-6d6a-11eb-89e3-eb8fa3995915.png)

These are just rough examples and are subject to balance changes on the sub builder's behalf. In this case I'm guessing it's TorpedoMike? (hi!!!)

----------------------------------------------------------------------------------------------------------------------------------------------

Another idea is the ability to further upgrade coilguns, railguns, and discharge coils. An example for the coilgun could be...

- A double barreled variant that fires at double the speed.
- A long barreled variant that deals more damage but fires slower. Uses more durability on coilgun ammo per shot?
- A chain gun variant that initially fires slowly, but quickly picks up speed. This windup could be reduced via upgrades?

An example for a railgun upgrade could be...

- A long barreled variant with a higher velocity and flatter trajectory. It would give more meaning to upgrading view range.
- An auto loader variant with quicker reload speed? This could potentially just be made into an increased reload speed upgrade.
- An experimental electric variant which charges the railgun shot and electrocutes any targets it hits. This could have a longer reload speed as a result.

An example for electrical discharge coils could be the ability to upgrade their range, damage, or stun times.

The general idea of this entire discussion is that the shipyard upgrade system can be built upon so that it allows players to upgrade their submarine on a more personalized basis. The ability to upgrade the weapon systems of the submarine is important in my eyes because the average player will often find that as they go deeper they will be dealing with creatures in both greater amounts and strength.

I also had a similar idea like this regarding the upgrading of handheld weapons such as the revolver, tactical shotgun, harpoon, and other various weapons at gunsmiths only found at military outposts. I'll save this for another discussion.

**Comment by Hinny** [comment]
Great idea! Would require a lot from sub builders, but would be give a lot of meaningful decisions.
